src/designspace.rs

```diff
@@ -168,55 +170,130 @@ impl DesignSpaceDocument {
         let reader = BufReader::new(File::open(path).map_err(DesignSpaceLoadError::Io)?);
         quick_xml::de::from_reader(reader).map_err(DesignSpaceLoadError::DeError)
     }
+
+    /// Save a designspace.
+    pub fn save(&self, path: impl AsRef<Path>) -> Result<(), DesignSpaceSaveError> {
+        let mut buf = String::from("<?xml version='1.0' encoding='UTF-8'?>\n");
+        let mut xml_writer = quick_xml::se::Serializer::new(&mut buf);
+        xml_writer.indent(' ', 2);
+        self.serialize(xml_writer)?;
+        buf.push('\n'); // trailing newline
+        fs::write(path, buf)?;
+        Ok(())
+    }
 }
 
 mod serde_impls {
 
     use super::{Axis, Dimension, Instance, Source};
-    use serde::{Deserialize, Deserializer};
-
-    pub fn deserialize_location<'de, D>(deserializer: D) -> Result<Vec<Dimension>, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        #[derive(Deserialize)]
-        struct Helper {
-            dimension: Vec<Dimension>,
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+    pub(super) mod location {
+        use super::*;
+
+        pub(crate) fn deserialize<'de, D>(deserializer: D) -> Result<Vec<Dimension>, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            #[derive(Deserialize)]
+            struct Helper {
+                dimension: Vec<Dimension>,
+            }
+            Helper::deserialize(deserializer).map(|x| x.dimension)
+        }
+
+        pub(crate) fn serialize<S>(location: &[Dimension], serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            #[derive(Serialize)]
+            struct Helper<'a> {
+                dimension: &'a [Dimension],
+            }
+            let helper = Helper { dimension: location };
+            helper.serialize(serializer)
         }
-        Helper::deserialize(deserializer).map(|x| x.dimension)
     }
 
-    pub fn deserialize_instances<'de, D>(deserializer: D) -> Result<Vec<Instance>, D::Error>
-    where
-        D: Deserializer<'de>,
-    {
-        #[derive(Deserialize)]
-        struct Helper {
-            instance: Vec<Instance>,
+    pub(super) mod instances {
+        use super::*;
+
+        pub(crate) fn deserialize<'de, D>(deserializer: D) -> Result<Vec<Instance>, D::Error>
+        where
+            D: Deserializer<'de>,
+        {
+            #[derive(Deserialize)]
+            struct Helper {
+                instance: Vec<Instance>,
+            }
+            Helper::deserialize(deserializer).map(|x| x.instance)
+        }
+
+        pub(crate) fn serialize<S>(instances: &[Instance], serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: Serializer,
+        {
+            #[derive(Serialize)]
+            struct Helper<'a> {
+                instance: &'a [Instance],
+            }
+            let helper = Helper { instance: instances };
+            helper.serialize(serializer)
```

Comment:
you know i bet we could have a simple `macro_rules` macro to handle all the duplicate code, but let's leave that as follow-up.